net/server.go

```diff
@@ -426,6 +419,116 @@ func (s *server) localEdges(tid thread.ID) (addrsEdge, headsEdge uint64, err err
 	return
 }
 
+// addPubSubTopic subscribes to a thread topic.
+func (s *server) addPubsubTopic(id thread.ID) error {
+	if s.ps == nil {
+		return nil
+	}
+	s.Lock()
+	defer s.Unlock()
+	if _, ok := s.topics[id]; ok {
+		return nil
+	}
+
+	t, err := rpc.NewTopic(s.net.ctx, s.ps, s.net.host.ID(), id.String(), true)
+	if err != nil {
+		return err
+	}
+	t.SetEventHandler(s.pubSubEventHandler)
+	t.SetMessageHandler(s.pubSubRecordHandler)
+	s.topics[id] = t
+	return nil
+}
+
+// addPubSubTopic subscribes to a thread topic.
+func (s *server) removePubsubTopic(id thread.ID) error {
+	if s.ps == nil {
+		return nil
+	}
+	s.Lock()
+	defer s.Unlock()
+	if t, ok := s.topics[id]; ok {
+		delete(s.topics, id)
+		return t.Close()
+	}
+	return nil
+}
+
+// addPubSubTopic subscribes to a thread topic.
+func (s *server) removeAllPubsubTopics() error {
+	if s.ps == nil {
+		return nil
+	}
+	s.Lock()
+	defer s.Unlock()
+	for id, t := range s.topics {
+		delete(s.topics, id)
+		if err := t.Close(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// publishRecord publishes a record request to a thread topic.
+func (s *server) publishRecord(ctx context.Context, topic thread.ID, req *pb.PushRecordRequest) error {
+	if s.ps == nil {
+		return nil
+	}
+	s.Lock()
+	t, ok := s.topics[topic]
+	s.Unlock()
+	if !ok {
+		return fmt.Errorf("publish to unknown thread %s", topic)
+	}
+
+	data, err := req.Marshal()
+	if err != nil {
+		return fmt.Errorf("marshaling pubsub message: %v", err)
+	}
+
+	if _, err := t.Publish(ctx, data, rpc.WithIgnoreResponse(true)); err != nil {
```

Comment:
With v0.0.4 of the rpc lib, each `Publish` call spawns a goroutine which keeps waiting for response until the context expires, even with this ignore response flag. It can cause too many goroutines if the timeout is long enough and `publishRecord` being called often. That might be an oversight, but depends on the answer to the question: if the caller doesn't expect a response, should `Publish` deliver the message to whichever peers online at that moment, or should it wait and re-publish to new peers until the context ends (same as the case of `WithMultiResponse`)?

Reply:
And we can change this behavior in the rpc lib pretty easily.

Reply:
hmm, good point! what do you think about an option in the rpc lib that is all about republishing: `rpc.WithRepublishing(true)` ?